### ISSUE TITLE
[To do] wvdecrypter Refactor

### ISSUE THREAD

**Issue by matthuisman** [peak3d/inputstream.adaptive#736]
We should move the wvdecrypter code into the main program as it's nearly always required.
This stops needing to load it up separately with dlopen.
It may speed it up, and makes it less complicated

Currently we have
ia.so > wvdecrypter.so > libwidevine.so
with above, we would have
ia.so > libwidevine.so

I think it started off separately as it needed to be shipped separate as the team didn't at that stage want the libwidevine integration code in the repo. that has now changed and currently wvdecrypter is shipped with all builds.

Maybe at the same time, the IA specific code for bento could be moved out of the bento files to allow for easier updating of the library.

**Comment by glennguy** [comment]
FYI libewidevinecdm is opened through `OpenDRMSystem` here https://github.com/xbmc/inputstream.adaptive/blob/260384fbf87c426fe40457ba41e770442ad31006/src/main.cpp#L2305

>it started off separately as it needed to be shipped separate

Yes but it's not necessarily a bad thing to not have it coupled to the main program. From my point of view it allows a future libssd_pr.so or libssd_fp.so to be used.

Check the `GetSupportedDecrypterURN` function in main.cpp and you can see that it's flexible to whichever key system you want to use. A decryptor can also support one or many key systems.

Another way to think of it - is this a solution in search of a problem?

Would love to hear @peak3d and/or @liberty-developer thoughts on this

**Comment by matthuisman** [comment]
@glennguy 
libwidevine is actually opened from the native_library files.
eg. https://github.com/xbmc/inputstream.adaptive/blob/Matrix/wvdecrypter/cdm/base/native_library_posix.cc#L25

i just think it'll simplify it a lot
No more needing to search for the decrypter for starters.
Surely it'll speed up loading it.
Might make debugging easier too?

If more decrypters are added, then just the same thing - have them built-in.
Not sure why they need to be separate binaries..
Especially when they are all shipped together anyway.

I'd just have them sit in a new "decrypter" directory in the inputstream.adaptive/src
eg. inputstream.adaptive/src/decrypter/widevine

I'm not too bothered either way, just seems like it would be a "nice to have"

**Comment by CastagnaIT** [comment]
if in future will be more type of derypters files, the loading time (and then playback starts/stop) could be worse?
probably have separate binaries is to limit the overall size of the main file, but we don't have such large sizes...

**Comment by glennguy** [comment]
@matthuisman I think this is still a 'to-do'.
After looking more since we last spoke of this it doesn't create tighter coupling in the code, and infact makes things much better in wvdecrypter side of things.
wvdecrypter has no access to the utilities in the kodi addon dev kit and there's no way to do this as far as I can see due to the way the build system is setup with the build_addon macro
But, if we bring the code into the main project then we can do `#include <kodi/Filesysten,h>` etc.
In other words I was wrong you were right :)

**Comment by matthuisman** [comment]
oh, i was just trying to clean up my issues :) i'll re-open

**Comment by tmm1** [comment]
I have mixed feelings about this. Having some separation is nice. Once you couple more tightly then it becomes harder to separate in the future.

I don't think there are performance concerns with the current setup. It's mostly about code cleanup and maintainability. Merging the code may make that simpler, but it also doesn't seem that bad right now.

**Comment by CastagnaIT** [comment]
why should you split all the code again? makes no sense

the main problem is not only maintenance, but all the workarounds applied around this, as:

unnecessary double codec interface very easy to break for external changes of which we are not aware (already happened recently),
workarounds applied for inability to use IA API interface from SSD project,
forcing the use of the InputStream API utils from the separate project (SSD) a code change that require a dependency may break it permanently,
external loading of another library that can be avoided,
inefficient code simplification or you hack all utils files to import them from other project or double all utils files thus creating more maintenance puzzles,

all this is a sign of poor bad software design of the whole current code

**Comment by glennguy** [comment]
What about having it both ways? I get the feeling that we're not all talking about the exact same thing here...

In my mind the biggest benefit comes from bringing the wvdecrypter project together with the main IA one in cmake, producing the one dll file. Then all Kodi functions and utils from the main project become available to use in wvdecrypter - unless there's another way to achieve this? It all seems heavily dependent on the `ADDONCREATOR` macro. 
We can still have the SSD interface as it is to keep things loosely coupled and have no references to the concrete classes from IA's side of the fence. Yes at the moment it seems pointless with one decrypter but I'd love to move forward with a clearkey one at some point in the near future and can appreciate the benefits of following this pattern.

**Comment by CastagnaIT** [comment]
to make it clearer to you what i think
i tried make a scheme
![immagine](https://user-images.githubusercontent.com/3257156/221409115-4adf133d-6644-42ec-bece-e2c05cb78c99.png)
